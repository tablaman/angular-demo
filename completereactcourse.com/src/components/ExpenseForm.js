import React from 'react';
import moment from 'moment';
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false
    }
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }
  onNoteChange = (e) => {
    // Note, need to do this - 
    // i.e. capture value initially
    // or use e.persist();
    const note = e.target.value;
    this.setState(() => ({ note }));
  }
  onAmountchange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
      this.setState(()=> ({ amount }))
  }

  // calendar
  onFocusChange = ({ focused }) => {
    this.setState(()=> ({ calendarFocused: focused }))
  }
  // better UX so user shouldn't be able to delete the date
  // from the input
  onDateChange = (createdAt) => {
    if (createdAt) this.setState(()=> ({ createdAt }));
  }
  onSubmit = (e) => {
    e.preventDefault();

    this.lblDescription.innerHTML = (!this.state.description) 
        ? 'please provide a description' : '';

    this.lblAmount.innerHTML = (!this.state.amount) 
        ? 'please provide an amount' : ''
    if(!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please fill out required fields' }))
    } else {
      this.setState(() => ({ error: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) *100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }
  render() {
    return (
      <div>
        { this.state.error && <p>{this.state.error}</p> }
        <form onSubmit={this.onSubmit}>
          <input 
            id="description"
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange} />
          <label 
            ref={el => this.lblDescription = el}
            htmlFor="description"
            ></label>

          <input 
            id="amount"
            type="text"
            placeholder="Amount"
            value = {this.state.amount}
            onChange={this.onAmountchange} />
          <label 
            ref={el => this.lblAmount = el}
            htmlFor="amount"
            ></label>

          <SingleDatePicker
            date={this.state.createdAt} 
            onDateChange={this.onDateChange} 
            focused={this.state.calendarFocused} 
            onFocusChange={this.onFocusChange} 
            numberOfMonths={1}
            isOutsideRange={() => false} // make every day available to us!
          />
          <textarea placeholder="Add a note for your expenses" 
            cols="30" rows="10"
            onChange={this.onNoteChange}></textarea>


          <button>Add expense</button>
        </form> 
      </div>
    )
  }
}

export default ExpenseForm;