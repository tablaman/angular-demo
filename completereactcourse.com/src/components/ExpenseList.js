import React from 'react';
import { connect } from 'react-redux';
import ExpenListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    <ul>
      {props.expenses.map(item => <ExpenListItem key={item.id} {...item} />)}
    </ul>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}
export default connect(mapStateToProps)(ExpenseList);