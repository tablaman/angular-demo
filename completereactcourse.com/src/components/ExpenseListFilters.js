import React from "react";
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate } from "../actions/filters";


const ExpenseListFilters = (props) => (
  <div>
    <input type="text" 
      value={props.filters.text}
      onChange={(e) => {
        props.dispatch(setTextFilter({text: e.target.value}));
      }} />
      <select 
        value={props.filters.sortBy}
        onChange={(e) => {
          switch(e.target.value) {
            case 'date':
              return props.dispatch(sortByDate())
            case 'amount':
              return props.dispatch(sortByAmount())
            default:
              break;
          }
        
      }}>
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
  </div>
);

const mapStateToProps = (state) => {
  return {
    filters: state.filters
  }
}


export default connect(mapStateToProps)(ExpenseListFilters);
