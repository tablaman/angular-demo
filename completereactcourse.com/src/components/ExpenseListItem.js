import React from 'react';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => (
  <li>
    <h3>
      <Link to={`/edit/${id}`}>{description} </Link>
    </h3>
    <p>{amount} - {createdAt}</p>
    <span>{note}</span>
  </li>
);

export default ExpenseListItem;