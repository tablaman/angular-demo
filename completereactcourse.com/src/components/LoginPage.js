import React from "react";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";
import { startLogin } from "../actions/auth";

export const LoginPage = ({startLogin}) => (
  <div>
      LOGIN PAGE
      <h1>Login</h1>
      <button onClick={ startLogin } >
        Login
      </button>
  </div>
)


const mapDispatchToProps = dispatch => ({
  startLogin: () => dispatch(startLogin())
});
// const mapDispatchToProps = dispatch => ({
//   startAddExpense: expense => dispatch(startAddExpense(expense))
// });

export default connect(undefined, mapDispatchToProps)(LoginPage);

// for testing purposes this has been altered
// typically we would use: export default connect()(AddExpensePage);
