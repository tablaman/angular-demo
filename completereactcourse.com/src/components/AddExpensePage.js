import React from 'react'
import ExpenseForm from './ExpenseForm';

const AddExpensePage = () => (
  <div>Add expense page
    <h1>Add Expense</h1>
    <ExpenseForm />
  </div>
  
)

export default AddExpensePage;