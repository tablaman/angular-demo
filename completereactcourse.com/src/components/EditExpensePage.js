// Edit Expense Page
import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  handleOnSubmit = expense => {
    this.props.editExpense({
      id: this.props.expense.id,
      updates: expense
    });
    this.props.history.push("/");
  }
  handleRemoveExpense = expense => {
    this.props.removeExpense({
      id: this.props.expense.id
    });
    this.props.history.push("/");
  }
  render () {
    return (
      <div>
        Edit expense page
        <Link to="/create">Create something</Link>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.handleOnSubmit}
        />
        <button
          onClick={this.handleRemoveExpense}
        >
          Remove
        </button>
      </div>

    )
  }
}
const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
