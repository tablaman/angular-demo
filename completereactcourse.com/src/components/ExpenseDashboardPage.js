//  Expense Dashboard Page
import React from 'react'
import {
  Link
} from 'react-router-dom'
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = (props) => (
  
  <div> Dashboard: expense page
    <Link to="/create">Create something</Link>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default ExpenseDashboardPage;