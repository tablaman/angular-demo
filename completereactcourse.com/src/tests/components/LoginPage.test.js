import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';


test('render ExpenseListItem with fixture data', () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot();
});