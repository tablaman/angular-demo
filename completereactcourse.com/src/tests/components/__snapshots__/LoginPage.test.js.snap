// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpenseListItem with fixture data 1`] = `
<div>
  LOGIN PAGE
  <h1>
    Login
  </h1>
  <button
    onClick={[Function]}
  />
</div>
`;
